fetch.js

```diff
@@ -459,6 +488,17 @@
         xhr.setRequestHeader(name, value)
       })
 
+      if (request.signal) {
+        request.signal.addEventListener('abort', abortXhr)
+
+        xhr.onreadystatechange = function() {
```

Comment:
this is the cleanest/safest way to know when the request has finished, regardless of the result, instead of having to do so on every callback:

```js
xhr.onload = function() {
  ...
  request.signal.removeEventListener('abort', abortXhr)
}

xhr.onabort = function() {
  ...
  request.signal.removeEventListener('abort', abortXhr)
}

xhr.onerror = function() {
  ...
  request.signal.removeEventListener('abort', abortXhr)
}

xhr.onprogress = function() {
  ...
  // should I remove listener? no.
}

// what other events there are? timeout?
...
```

having the `removeListener` close to `addListener` is also easier to read/maintain.

Reply:
Another possibility is [`xhr.onloadend`](https://xhr.spec.whatwg.org/#event-xhr-loadend). This event is dispatched after the request completes, regardless of *how* completes (`load`, `error`, `abort` or `timeout`).